### PR TITLE
feat(LogPerf): collect XSLogs to SimTop.LogPerfEndpoint

### PR DESCRIPTION
**Pull request by klin02**
XSLog depends on LogPerfCtrl declared at Top Module. Previous we annotate such signal as dontTouch, and accessed through Hierarchical name like SimTop.xx by dummy LogPerfHelper.

However, as XSLog is called in many spaces in DUT, which are not visible to each other, especailly some in WhenContext. XS will generate thousands of LogPerfHelper to get same LogPerfCtrl. Too many module instantiations greately slow down compilation speed, especailly in Palladium (more than 5 times slower than same DUT without Log).

This change collect all XSLog to SimTop.LogPerfEndpoint, with LogPerfCtrl directly passed by IO. Some tips as follows:

1. Not call XSLog inside whenContext. To collect XSLogs, we should access Cond and Data from other module, but data in WhenContext is not accessible even through tap. Use XSLog(cond, pable) instead of when(cond) {XSLog(pable)}. We also add chisel Internal API `currentWhen` to check that.

2. Generate Hierarchical Module path through FIRRTL transforms. Sometimes we want to append module path for better debugging. XSCompatibility add a hacky way to use Chisel internal API to get tag of current Module. Then we will replace these tag with path during ChiselStage. Note path can only be acessed after circuit elaboration.

3. Register and invoke caller of XSPerf and related object. As XSPerf depends on LogPerfCtrl such as dump. We should deferred apply() until collect. So we regirster collect() method when firstly apply XSLog, then XSLog will automatically call XSPerf.collect() method during collection. Note deferred apply is called in another module, so original module tag should be recorded for path generation.

### CODE REVIEW

**Comment by klin02** [comment]
Still a draft. Code format and some other logic need to be improved later.

**Comment by klin02** [comment]
Emitted example
```
if (SimTop.l_soc.core_with_l2.core.frontend.inner_bpu.predictors.tage.tables_0.bore) $fwrite(32'h80000002, "[ERROR][time=%d] SimTop.l_soc.core_with_l2.core.frontend.inner_bpu.predictors.tage.tables_0: tage table 0 has different fh, ghist: %b, fh: %b\n", io_timer, SimTop.l_soc.core_with_l2.core.frontend.inner_bpu.predictors.tage.tables_0.data_0, SimTop.l_soc.core_with_l2.core.frontend.inner_bpu.predictors.tage.tables_0.data_1);
if (SimTop.endpoint.bore && io_logEnable) $fwrite(32'h80000002, "[PERF ][time=%d] SimTop.endpoint: L3_Mem_L3_bank_0_A_channel_fire, %d\n", io_timer, SimTop.endpoint.data_0);
```

**Comment by klin02** [comment]
If we can ensure XSLog outside whenContext(also needed by emitting directly) , all data can be accessed by tap. Then we only need to get hierarchical module name after circuit elaboration. A more pretty way is to generate LogMacros.svh after elaboration and add "`include LogMacros` to generated LogPerfEndpoint.

**Comment by klin02** [comment]
Done. And update first comment with commitMessage

**Review by poemonsense — APPROVED** [review]
The code itself looks good to me, as I believe you have tested it on XiangShan.

However, I'm worried on how we maintain the code. The logs depend on top-level IOs, and these IOs are written in either DiffTest or other simulation frameworks than this Utility repository. Now this PR seems to introduce more Chisel APIs as well. Through separating different repository, we expect that difference repos can have their own versions and developments. However, as we are continuing mixing things together in some repositories, the boundaries of different repositories seem unclear.

I don't have a good specific idea on how we should maintain the code. But I think there are at least something we can do to improve it.

1) If we do need the currentModule and currentWhen, we can help improve Chisel APIs. For the latter, it seems this PR only relies on the existence of When, instead of the actual WhenContext.

2) If we do need to access all printf's after elaboration, we can propose this to Chisel as well. I would prefer calling it "lazy printf" as we are not creating it during elaboration but delaying it to an explicit instantiation call.

3) We may think of a meta API that the user could use for performance counters. This API may not depend on the simulation framework interfaces. Instead, they can manage the performance counters, letting the user to decide whether these counters are displayed via printf, lazy printf, or any other ways.

@sequencer @Tang-Haojin Any ideas on managing XiangShan (third party designs, Chisel users), util (base libraries), Chisel, etc code better? If Chisel dev thinks it's good, we can try upstream some of the code. In such ways, at least we are driving the Chisel/open-source community forward, instead of maintaining some in-house code that no one else than XiangShan dev would ever use.

**Comment by sequencer** [comment]
I personally prefer not using `tapAndRead` but explicitly using probe api for module level interface stability. But this feature is maintained in Chisel, and you can confidentially use it.

**Comment by sequencer** [comment]
FWIW, I think you can take a look at T1’s verification framework. It also use the differential test framework but totally based on the output of the simulation. It also doesn’t tap internal signals but exposes all necessary with Probe and Layer API.
We need to keep these verification logic out of our customers RTL to reduce the delivery aspect.

**Comment by poemonsense** [comment]
> FWIW, I think you can take a look at T1’s verification framework. It also use the differential test framework but totally based on the output of the simulation. It also doesn’t tap internal signals but exposes all necessary with Probe and Layer API. We need to keep these verification logic out of our customers RTL to reduce the delivery aspect.

Thanks for the information.

@klin02 Let's first merge this as the initial v1 version after @Tang-Haojin approves it. We can discuss the improvements in the next steps.

**Comment by klin02** [comment]
OK. And later I will take a look at T1 practice for better framework.

**Comment by Tang-Haojin** [comment]
Generally LGTM. Have you tested the impact of elaboration, compilation and simulation speed? How many threads can we use for a verilated model with this midification?

**Comment by klin02** [comment]
> Generally LGTM. Have you tested the impact of elaboration, compilation and simulation speed? How many threads can we use for a verilated model with this modification?

I will make more test before merging it to ensure cmopilation and simulation speed.

**Comment by klin02** [comment]
Previous we apply printf to each printable, too many fwrites make it hard to schedule, causing UNOPTTHREADS warning with 16-threads Verilator.

Consider many XSLogs may have same printf condition(especially XSPerfs, takes up a large amount), we group XSLogs with same condition, and concat their Printables (Also, limit size less than 1000 to avoid segment fault caused by too long printf）. After this change, compilation passed with 16-threads Verilator.

**Review by Tang-Haojin — APPROVED** [review]

**Pull request merged**
